File: js/managers/SettingsManager.js
/**
 * SettingsManager - Manages all app settings with localStorage persistence
 * Direct port of Swift SettingsManager
 */
import { eventBus, Events } from '../core/EventBus.js';

const STORAGE_KEY = 'stopMotionSettings';

const DEFAULT_SETTINGS = {
    // Recognition
    recognitionType: 'qrCode', // 'qrCode' | 'barcode' | 'colorSample'
    captureDelay: 0.5, // seconds

    // Color targeting
    targetColor: { r: 1.0, g: 0.0, b: 0.0 }, // RGB 0-1

    // Frame overlay
    frameTopThickness: 80, // pixels
    frameBottomThickness: 80, // pixels

    // Playback
    frameRate: 12, // fps
    reverseMovie: false,

    // Camera (limited in web)
    zoomFactor: 1.3,

    // Effects
    kaleidoscopeEnabled: false,
    kaleidoscopeRotation: 0,

    // Onion skinning
    onionSkinEnabled: true,
    onionSkinOpacity: 0.6,
    onionSkinOffsetX: 0,
    onionSkinOffsetY: 0,

    // Image adjustments
    contrast: 1.0,
    saturation: 1.0,
    invertColors: false,
    threshold: 0,

    // Transparency
    backgroundTransparent: false,
    transparencyAdjust: 0,
    useBackground: false,

    // Network
    serverAddress: '192.168.1.198:8080'
};

class SettingsManagerClass {
    constructor() {
        this._settings = { ...DEFAULT_SETTINGS };
        this._listeners = new Map();
        this.load();
    }

    /**
     * Load settings from localStorage
     */
    load() {
        try {
            const stored = localStorage.getItem(STORAGE_KEY);
            if (stored) {
                const parsed = JSON.parse(stored);
                this._settings = { ...DEFAULT_SETTINGS, ...parsed };
            }
        } catch (error) {
            console.warn('Failed to load settings:', error);
            this._settings = { ...DEFAULT_SETTINGS };
        }
    }

    /**
     * Save settings to localStorage
     */
    save() {
        try {
            localStorage.setItem(STORAGE_KEY, JSON.stringify(this._settings));
        } catch (error) {
            console.warn('Failed to save settings:', error);
        }
    }

    /**
     * Reset all settings to defaults
     */
    reset() {
        this._settings = { ...DEFAULT_SETTINGS };
        this.save();
        eventBus.publish(Events.SETTINGS_CHANGED, { all: true });
    }

    /**
     * Get a setting value
     * @param {string} key - Setting key
     * @returns {*} Setting value
     */
    get(key) {
        return this._settings[key];
    }

    /**
     * Set a setting value
     * @param {string} key - Setting key
     * @param {*} value - New value
     */
    set(key, value) {
        if (this._settings[key] !== value) {
            this._settings[key] = value;
            this.save();

            // Notify listeners
            const listeners = this._listeners.get(key);
            if (listeners) {
                listeners.forEach(callback => callback(value));
            }

            eventBus.publish(Events.SETTINGS_CHANGED, { key, value });
        }
    }

    /**
     * Subscribe to changes for a specific setting
     * @param {string} key - Setting key
     * @param {Function} callback - Handler function
     * @returns {Function} Unsubscribe function
     */
    subscribe(key, callback) {
        if (!this._listeners.has(key)) {
            this._listeners.set(key, new Set());
        }
        this._listeners.get(key).add(callback);

        return () => {
            const listeners = this._listeners.get(key);
            if (listeners) {
                listeners.delete(callback);
            }
        };
    }

    /**
     * Get all settings as an object
     * @returns {Object} All settings
     */
    getAll() {
        return { ...this._settings };
    }

    // Convenience getters for common settings
    get recognitionType() { return this._settings.recognitionType; }
    set recognitionType(v) { this.set('recognitionType', v); }

    get captureDelay() { return this._settings.captureDelay; }
    set captureDelay(v) { this.set('captureDelay', v); }

    get targetColor() { return this._settings.targetColor; }
    set targetColor(v) { this.set('targetColor', v); }

    get frameTopThickness() { return this._settings.frameTopThickness; }
    set frameTopThickness(v) { this.set('frameTopThickness', v); }

    get frameBottomThickness() { return this._settings.frameBottomThickness; }
    set frameBottomThickness(v) { this.set('frameBottomThickness', v); }

    get frameRate() { return this._settings.frameRate; }
    set frameRate(v) { this.set('frameRate', v); }

    get reverseMovie() { return this._settings.reverseMovie; }
    set reverseMovie(v) { this.set('reverseMovie', v); }

    get zoomFactor() { return this._settings.zoomFactor; }
    set zoomFactor(v) { this.set('zoomFactor', v); }

    get kaleidoscopeEnabled() { return this._settings.kaleidoscopeEnabled; }
    set kaleidoscopeEnabled(v) { this.set('kaleidoscopeEnabled', v); }

    get kaleidoscopeRotation() { return this._settings.kaleidoscopeRotation; }
    set kaleidoscopeRotation(v) { this.set('kaleidoscopeRotation', v); }

    get onionSkinEnabled() { return this._settings.onionSkinEnabled; }
    set onionSkinEnabled(v) { this.set('onionSkinEnabled', v); }

    get onionSkinOpacity() { return this._settings.onionSkinOpacity; }
    set onionSkinOpacity(v) { this.set('onionSkinOpacity', v); }

    get onionSkinOffsetX() { return this._settings.onionSkinOffsetX; }
    set onionSkinOffsetX(v) { this.set('onionSkinOffsetX', v); }

    get onionSkinOffsetY() { return this._settings.onionSkinOffsetY; }
    set onionSkinOffsetY(v) { this.set('onionSkinOffsetY', v); }

    get contrast() { return this._settings.contrast; }
    set contrast(v) { this.set('contrast', v); }

    get saturation() { return this._settings.saturation; }
    set saturation(v) { this.set('saturation', v); }

    get invertColors() { return this._settings.invertColors; }
    set invertColors(v) { this.set('invertColors', v); }

    get threshold() { return this._settings.threshold; }
    set threshold(v) { this.set('threshold', v); }

    get backgroundTransparent() { return this._settings.backgroundTransparent; }
    set backgroundTransparent(v) { this.set('backgroundTransparent', v); }

    get transparencyAdjust() { return this._settings.transparencyAdjust; }
    set transparencyAdjust(v) { this.set('transparencyAdjust', v); }

    get useBackground() { return this._settings.useBackground; }
    set useBackground(v) { this.set('useBackground', v); }

    get serverAddress() { return this._settings.serverAddress; }
    set serverAddress(v) { this.set('serverAddress', v); }
}

// Singleton instance
export const settings = new SettingsManagerClass();

File: Overhead_iPhone_Animations/Overhead_iPhone_Animations/SettingsManager.swift
//
//  SettingsManager.swift
//  ChristmasStopFrame
//
//  Created by Durrell Bishop on 21/12/2025.
//

import Foundation
import Combine

class SettingsManager: ObservableObject {
    static let shared = SettingsManager()

    private enum Keys {
        static let recognitionType = "recognitionType"
        static let captureDelay = "captureDelay"
        static let targetColorRed = "targetColorRed"
        static let targetColorGreen = "targetColorGreen"
        static let targetColorBlue = "targetColorBlue"
        static let frameTopThickness = "frameTopThickness"
        static let frameBottomThickness = "frameBottomThickness"
        static let frameRate = "frameRate"
        static let zoomFactor = "zoomFactor"
        static let kaleidoscopeEnabled = "kaleidoscopeEnabled"
        static let serverAddress = "serverAddress"
        static let onionSkinEnabled = "onionSkinEnabled"
        static let onionSkinOffsetX = "onionSkinOffsetX"
        static let onionSkinOffsetY = "onionSkinOffsetY"
        static let onionSkinOpacity = "onionSkinOpacity"
        static let exposureBias = "exposureBias"
        static let contrast = "contrast"
        static let saturation = "saturation"
        static let invertColors = "invertColors"
        static let threshold = "threshold"
        static let backgroundTransparent = "backgroundTransparent"
        static let transparencyAdjust = "transparencyAdjust"
        static let useBackground = "useBackground"
        static let reverseMovie = "reverseMovie"
    }

    @Published var recognitionType: RecognitionType {
        didSet {
            UserDefaults.standard.set(recognitionType.rawValue, forKey: Keys.recognitionType)
        }
    }

    @Published var captureDelay: Double {
        didSet {
            UserDefaults.standard.set(captureDelay, forKey: Keys.captureDelay)
        }
    }

    @Published var targetColor: (red: Double, green: Double, blue: Double) {
        didSet {
            UserDefaults.standard.set(targetColor.red, forKey: Keys.targetColorRed)
            UserDefaults.standard.set(targetColor.green, forKey: Keys.targetColorGreen)
            UserDefaults.standard.set(targetColor.blue, forKey: Keys.targetColorBlue)
        }
    }

    @Published var frameTopThickness: Double {
        didSet {
            UserDefaults.standard.set(frameTopThickness, forKey: Keys.frameTopThickness)
        }
    }

    @Published var frameBottomThickness: Double {
        didSet {
            UserDefaults.standard.set(frameBottomThickness, forKey: Keys.frameBottomThickness)
        }
    }

    @Published var frameRate: Double {
        didSet {
            UserDefaults.standard.set(frameRate, forKey: Keys.frameRate)
        }
    }

    @Published var zoomFactor: Double {
        didSet {
            UserDefaults.standard.set(zoomFactor, forKey: Keys.zoomFactor)
        }
    }

    @Published var kaleidoscopeEnabled: Bool {
        didSet {
            UserDefaults.standard.set(kaleidoscopeEnabled, forKey: Keys.kaleidoscopeEnabled)
        }
    }

    @Published var serverAddress: String {
        didSet {
            UserDefaults.standard.set(serverAddress, forKey: Keys.serverAddress)
        }
    }

    @Published var onionSkinEnabled: Bool {
        didSet {
            UserDefaults.standard.set(onionSkinEnabled, forKey: Keys.onionSkinEnabled)
        }
    }

    @Published var onionSkinOffsetX: Double {
        didSet {
            UserDefaults.standard.set(onionSkinOffsetX, forKey: Keys.onionSkinOffsetX)
        }
    }

    @Published var onionSkinOffsetY: Double {
        didSet {
            UserDefaults.standard.set(onionSkinOffsetY, forKey: Keys.onionSkinOffsetY)
        }
    }

    @Published var onionSkinOpacity: Double {
        didSet {
            UserDefaults.standard.set(onionSkinOpacity, forKey: Keys.onionSkinOpacity)
        }
    }

    @Published var exposureBias: Double {
        didSet {
            UserDefaults.standard.set(exposureBias, forKey: Keys.exposureBias)
        }
    }

    @Published var contrast: Double {
        didSet {
            UserDefaults.standard.set(contrast, forKey: Keys.contrast)
        }
    }

    @Published var saturation: Double {
        didSet {
            UserDefaults.standard.set(saturation, forKey: Keys.saturation)
        }
    }

    @Published var invertColors: Bool {
        didSet {
            UserDefaults.standard.set(invertColors, forKey: Keys.invertColors)
        }
    }

    @Published var threshold: Double {
        didSet {
            UserDefaults.standard.set(threshold, forKey: Keys.threshold)
        }
    }

    @Published var backgroundTransparent: Bool {
        didSet {
            UserDefaults.standard.set(backgroundTransparent, forKey: Keys.backgroundTransparent)
        }
    }

    @Published var transparencyAdjust: Double {
        didSet {
            UserDefaults.standard.set(transparencyAdjust, forKey: Keys.transparencyAdjust)
        }
    }

    @Published var useBackground: Bool {
        didSet {
            UserDefaults.standard.set(useBackground, forKey: Keys.useBackground)
        }
    }

    @Published var reverseMovie: Bool {
        didSet {
            UserDefaults.standard.set(reverseMovie, forKey: Keys.reverseMovie)
        }
    }

    private init() {
        let savedTypeString = UserDefaults.standard.string(forKey: Keys.recognitionType) ?? RecognitionType.qrCode.rawValue
        self.recognitionType = RecognitionType(rawValue: savedTypeString) ?? .qrCode

        self.captureDelay = UserDefaults.standard.object(forKey: Keys.captureDelay) as? Double ?? 0.5

        let red = UserDefaults.standard.object(forKey: Keys.targetColorRed) as? Double ?? 1.0
        let green = UserDefaults.standard.object(forKey: Keys.targetColorGreen) as? Double ?? 0.0
        let blue = UserDefaults.standard.object(forKey: Keys.targetColorBlue) as? Double ?? 0.0
        self.targetColor = (red, green, blue)

        self.frameTopThickness = UserDefaults.standard.object(forKey: Keys.frameTopThickness) as? Double ?? 80.0
        self.frameBottomThickness = UserDefaults.standard.object(forKey: Keys.frameBottomThickness) as? Double ?? 80.0
        self.frameRate = UserDefaults.standard.object(forKey: Keys.frameRate) as? Double ?? 12.0
        self.zoomFactor = UserDefaults.standard.object(forKey: Keys.zoomFactor) as? Double ?? 1.3
        self.kaleidoscopeEnabled = UserDefaults.standard.bool(forKey: Keys.kaleidoscopeEnabled)
        self.serverAddress = UserDefaults.standard.string(forKey: Keys.serverAddress) ?? "192.168.1.198:8080"
        self.onionSkinEnabled = UserDefaults.standard.object(forKey: Keys.onionSkinEnabled) as? Bool ?? true
        self.onionSkinOffsetX = UserDefaults.standard.object(forKey: Keys.onionSkinOffsetX) as? Double ?? 0.0
        self.onionSkinOffsetY = UserDefaults.standard.object(forKey: Keys.onionSkinOffsetY) as? Double ?? 0.0
        self.onionSkinOpacity = UserDefaults.standard.object(forKey: Keys.onionSkinOpacity) as? Double ?? 0.6
        self.exposureBias = UserDefaults.standard.object(forKey: Keys.exposureBias) as? Double ?? 0.0
        self.contrast = UserDefaults.standard.object(forKey: Keys.contrast) as? Double ?? 1.0
        self.saturation = UserDefaults.standard.object(forKey: Keys.saturation) as? Double ?? 1.0
        self.invertColors = UserDefaults.standard.bool(forKey: Keys.invertColors)
        self.threshold = UserDefaults.standard.object(forKey: Keys.threshold) as? Double ?? 0.0
        self.backgroundTransparent = UserDefaults.standard.bool(forKey: Keys.backgroundTransparent)
        self.transparencyAdjust = UserDefaults.standard.object(forKey: Keys.transparencyAdjust) as? Double ?? 0.0
        self.useBackground = UserDefaults.standard.bool(forKey: Keys.useBackground)
        self.reverseMovie = UserDefaults.standard.bool(forKey: Keys.reverseMovie)
    }
}
